actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -467,15 +469,29 @@ def initialize(parent_jar)
 
       def [](name)
         if data = @parent_jar[name.to_s]
-          parse name, data
+          value = parse name, data
+          if Cookies.cookies_have_purpose
+            if value["purpose"].eql?(name.to_s)
```

Comment:
This isn't exactly what we mean by purpose, though defaulting to verifying that the name fits can be a good default!
Consider that a `:user_id` cookie can be used for different purposes and then we do need to provide a way to specify it.

Some possible APIs:

```ruby
cookies.signed.purpose(:login)[:user_id]

cookies.signed.read(:user_id, for: :login)
cookies.signed.write(:user_id, for: :voicemails, expires_in: 2.hours) # To separate a users general login from the auth to listen to voicemails.
```